### ISSUE TITLE
Use of regex! involves .so file, which executable does not find.

### ISSUE THREAD

**Issue by John-Nagle** [rust-lang/rust#21861]
(Using rustc 1.0.0-nightly (1d00c545e 2015-01-30 19:56:34 +0000)) 

Program using an external crate which uses "regex!" builds normally, but upon execution: 
"target/rss_client: error while loading shared libraries: libregex_macros-fdf5a2d7af19fd78.so: cannot open shared object file: No such file or directory"

The desired .so file is in target/deps, and the executable doesn't look there unless I set LD_LIBRARY_PATH.  Then the program runs.

But why is "regex" in a .so file? Here's my target/deps directory:

    libchrono-2760812556f7a811.rlib
    libcookie-4a08a2c06bb5c4fc.rlib
    libgcc-3361713823cacea0.rlib
    libhyper-9e6a74d72671a096.rlib
    liblibc-6b537f2898b2a028.rlib
    liblog-b7487854e865255d.rlib
    libmatches-68db25b520030534.rlib
    libmime-115ca8125c42dc12.rlib
    libmucell-178e4fc583386697.rlib
    libopenssl-ff2380beb920b0db.rlib
    libopenssl-sys-1d498ee6f4e8d500.rlib
    libpkg-config-be6313c85913bcf0.rlib
    libregex-50925f1e98b93309.rlib
    libregex_macros-fdf5a2d7af19fd78.so <---- How did that get in there?
    librustc-serialize-181d5959361aa206.rlib
    libtime-2215e3ac75eaa2c7.rlib
    libunicase-88de2458e97b894d.rlib
    libunsafe-any-3f018797a3613438.rlib
    liburl-4c384fb33b0bde93.rlib
    libxml-25cb934bf93db2fd.rlib

Everything else is statically linked, and normally nothing in the "deps" directory is needed for execution. Why did "cargo build" generate a ".so" file? Did someone build "regex" as a dynamic library by mistake? I've been looking at the relevant .toml files for something that would cause this, but I'm not finding it.

Where is static vs. dynamic library generation under Cargo control documented, anyway? Not finding anything in "http://doc.crates.io/guide.html". Obviously there's something somewhere which says "make this a shared library". But where?

Again, the program works if LD_LIBRARY_PATH includes the directory with that .so file.

**Comment by steveklabnik** [comment]
> Where is static vs. dynamic library generation under Cargo control
documented, anyway?

I've been meaning to consolidate all of Cargo's documentation, but haven't
had the chance. It's here:
http://doc.crates.io/manifest.html#building-dynamic-or-static-libraries

**Comment by John-Nagle** [comment]
OK. The relevant dependences are in the crate which uses "regex" (https://github.com/John-Nagle/rust-chrono)  which is the "chrono" package with some parsing functions added. The versions used are:

    regex = "0.1.12"
    regex_macros = "0.1.6"

Relevant .toml files of those versions:
    https://github.com/rust-lang/regex/blob/6c0f8aa82582441e9cb3813048ef9b098c026db0/regex_macros/Cargo.toml
    https://github.com/rust-lang/regex/blob/master/Cargo.toml

The relevant section of the regex_macros .toml file:

    [lib]
    name = "regex_macros"
    plugin = true

The unusual things about the regex_macros crate are that it's a crate within a crate (it's a subtree of regex) and it's a plugin. Could either of those trigger use as a .so file?

**Comment by huonw** [comment]
Plugins are dynamic libraries (they're dynamically loaded into the compiler), they should not be needed at runtime at all. The correct way to load them is to use the `#[no_link]` attribute, it seems that chrono has this, maybe something is loading `regex_macros` as a plugin without `no_link`?

**Comment by alexcrichton** [comment]
Yes as @huonw mentioned this looks like intended behavior, so I'm going to close this.

@John-Nagle do you have a `Cargo.lock` I could look at? I might be able to help out in tracking down which of your deps forgot the `#[no_link]` tag.

**Comment by alexcrichton** [comment]
Yes as @huonw mentioned this looks like intended behavior, so I'm going to close this.

@John-Nagle do you have a `Cargo.lock` I could look at? I might be able to help out in tracking down which of your deps forgot the `#[no_link]` tag.

**Comment by John-Nagle** [comment]
OK. So somewhere, something supposedly pulled in "regex_macros" without #[no_link]. I cannot find where. The program giving trouble is my "rss_client". It does not use the regex package directly, and never mentions it, but uses my version of "rust_chrono", which uses regex for date-parsing functions. 

I can't find any place outside chrono that mentions regex at all. Nor is there a "macro_reexport" which would push the macros another level outward. I may be doing something wrong, but if so, it's probably something that isn't in the manual.

Miscellaneous fact: I have Rust installed in ~/local, not installed as root. Ubuntu 14.04.

Also, "cargo test" works fine, presumably because Cargo sets LD_LIBRARY_PATH to include the "target/deps" directory.

Here's the rss_client tomo.lock file: 

```[root]
name = "rss_client"
version = "0.0.1"
dependencies = [
 "RustyXML 0.0.1 (git+https://github.com/Florob/RustyXML)",
 "chrono 0.1.15 (git+https://github.com/John-Nagle/rust-chrono)",
 "hyper 0.1.9 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "RustyXML"
version = "0.0.1"
source = "git+https://github.com/Florob/RustyXML#2072b4687faf659e018e7d3ee2d9e1eff5b345f8"

[[package]]
name = "chrono"
version = "0.1.15"
source = "git+https://github.com/John-Nagle/rust-chrono#a8c866d1ea08a2e5613bd5811f146b7c2e5cf987"
dependencies = [
 "regex 0.1.12 (registry+https://github.com/rust-lang/crates.io-index)",
 "regex_macros 0.1.6 (registry+https://github.com/rust-lang/crates.io-index)",
 "time 0.1.15 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "cookie"
version = "0.1.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "openssl 0.2.18 (registry+https://github.com/rust-lang/crates.io-index)",
 "rustc-serialize 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)",
 "time 0.1.15 (registry+https://github.com/rust-lang/crates.io-index)",
 "url 0.2.17 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "gcc"
version = "0.1.7"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "hyper"
version = "0.1.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "cookie 0.1.9 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
 "mime 0.0.7 (registry+https://github.com/rust-lang/crates.io-index)",
 "mucell 0.1.12 (registry+https://github.com/rust-lang/crates.io-index)",
 "openssl 0.2.18 (registry+https://github.com/rust-lang/crates.io-index)",
 "rustc-serialize 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)",
 "time 0.1.15 (registry+https://github.com/rust-lang/crates.io-index)",
 "unicase 0.0.3 (registry+https://github.com/rust-lang/crates.io-index)",
 "unsafe-any 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "url 0.2.17 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "libc"
version = "0.1.1"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "libressl-pnacl-sys"
version = "2.1.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "pnacl-build-helper 1.3.2 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "log"
version = "0.2.1"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "matches"
version = "0.1.2"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "mime"
version = "0.0.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "log 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "mucell"
version = "0.1.12"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "openssl"
version = "0.2.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "openssl-sys 0.2.18 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "openssl-sys"
version = "0.2.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "libressl-pnacl-sys 2.1.4 (registry+https://github.com/rust-lang/crates.io-index)",
 "pkg-config 0.1.7 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "pkg-config"
version = "0.1.7"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "pnacl-build-helper"
version = "1.3.2"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "regex"
version = "0.1.12"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "regex_macros"
version = "0.1.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "regex 0.1.12 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "rustc-serialize"
version = "0.2.10"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "time"
version = "0.1.15"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "gcc 0.1.7 (registry+https://github.com/rust-lang/crates.io-index)",
 "libc 0.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
]

[[package]]
name = "unicase"
version = "0.0.3"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "unsafe-any"
version = "0.2.2"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "url"
version = "0.2.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "matches 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "rustc-serialize 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)",
]```

**Comment by alexcrichton** [comment]
It looks like you may be accidentally [linking to `regex_macros` twice](https://github.com/John-Nagle/rust-chrono/blob/master/src/parse.rs#L1). The first one has `#[plugin]` and `#[no_link]`, but the second one does not.

**Comment by John-Nagle** [comment]
You may be right. But that's what the documentation says to do.  The example in the manual, at
http://doc.rust-lang.org/regex/regex/index.html
reads
```#![feature(plugin)]
#[plugin] #[no_link]
extern crate regex_macros;
extern crate regex;

fn main() {
    let re = regex!(r"^\d{4}-\d{2}-\d{2}$");
    assert_eq!(re.is_match("2014-01-01"), true);
}```

"regex_macros" is a crate within the "regex" crate. That's probably confusing something in the build/link system.

**Comment by alexcrichton** [comment]
Ah you already have the necessary lines [in `src/lib.rs`](https://github.com/John-Nagle/rust-chrono/blob/master/src/lib.rs#L195-L197) so they just don't need to be duplicated in `src/parse.rs`. I don't think you'll even need to import `regex_macros` in `src/parse.rs` to get the macros to work.

Note that the attributes in the documentation (`#[plugin] #[no_link]`) are crucial to get it to work. Also note that this is one of many reasons why plugins are currently unstable :)

**Comment by alexcrichton** [comment]
> There's a .toml file under another .toml file.

This is actually a feature of Cargo where one repository can host many crates by having a separate directory for each crate.

**Comment by John-Nagle** [comment]
Ah. That's it. I removed the imports in src/parse.rs, the dependency on the .so file disappeared from the final executable, and the program runs correctly. Thanks.

The compiler probably should detect such attribute inconsistencies across files. This sort of thing is likely to creep in as programs are refactored. That's how I got into this; the date/time parsing module was originally part of the RSS reader, and then I moved it to "chrono" with the rest of the date/time functions.
